### DISCUSSION TITLE
Dokku performance benchmark

### DISCUSSION THREAD

**Discussion by gal064** [dokku/dokku#5029] [General]
Hi all,

I wonder if anyone ever tested how (on average) using Dokku to deploy an app impact the performance of the app?

I know it depends on the app itself, the language, config etc - but just understanding the order of magnitude would be helpful.

For example, let's say I'd like to deploy a self-hosted Ghost app to a $5 VPS (e.g. Digitalocean). I have two options,  use Dokku or install Ghost + NGINX + MySQL. I wonder how using Dokku will impact performance

**Comment by josegonzalez** [comment]
Dokku starts your app in a docker container and routes to it via nginx. The overhead would be of running the container (there is some, though minimal).

I did some benchmarking 7 years ago (so these results should be taken with a grain of salt). Here is a link: https://github.com/dokku/dokku/issues/1576#issuecomment-150785023

There have definitely been changes in networking performance for Docker over the years. Example [here](https://jtway.co/docker-network-performance-b95bce32b4b9).

All that said, if you're hosting a blog on a single $5 VPS, the performance is probably going to be more than fine.

**Reply by gal064** [reply]
Thanks! I hear a some people say that Dokku is not the best option for large scale apps and recommend it for personal projects. If Dokku is mainly a wrapper around Docker, what's the reason to think it's not suitable for scale?

**Reply by StitiFatah** [reply]
"Scale" should be defined but probably because of lack of horizontal scaling, but I bet you could go pretty far with vertical scaling alone.

**Reply by josegonzalez** [reply]
You can scale horizontally with our Kubernetes plugin.

Again, the overhead here comes from the docker container stack and whatever routing layer you have in front. Nginx has been proven to be fairly high-performance, and while our default settings are pretty good, there is probably always something you can do to make it better for your specific use-case. Docker's networking layer and Linux container technologies in general have _some_ overhead that you won't be able to overcome, but I think the benefits of the system - easily replicable deploys - outweigh any drawbacks.

Finally, for 90% of our users, your time is better spent resizing your server to be 1 size bigger (whatever that means) than trying to squeeze out an extra 5% performance.
